sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -1208,28 +1208,34 @@ private static bool TryGetValueBody(AmqpMessage source, out AmqpMessageBody valu
         ///
         private static ArraySegment<byte> ReadStreamToArraySegment(Stream stream)
         {
-            if (stream == null)
-            {
-                return new ArraySegment<byte>();
-            }
-
             switch (stream)
             {
+                case { Length: < 1 }:
+                    return default;
+
                 case BufferListStream bufferListStream:
                     return bufferListStream.ReadBytes((int)stream.Length);
 
                 case MemoryStream memStreamSource:
                 {
                     using var memStreamCopy = new MemoryStream((int)(memStreamSource.Length - memStreamSource.Position));
                     memStreamSource.CopyTo(memStreamCopy, StreamBufferSizeInBytes);
-                    return new ArraySegment<byte>(memStreamCopy.ToArray());
+                    if (!memStreamCopy.TryGetBuffer(out ArraySegment<byte> segment))
+                    {
+                        segment = new ArraySegment<byte>(memStreamCopy.ToArray());
+                    }
+                    return segment;
                 }
```

Comment:
```suggestion
                    if (!memStreamSource.TryGetBuffer(out ArraySegment<byte> segment))
                    {
                        using var memStreamCopy = new MemoryStream((int)(memStreamSource.Length - memStreamSource.Position));
                        memStreamSource.CopyTo(memStreamCopy, StreamBufferSizeInBytes);
                        
                        segment = new ArraySegment<byte>(memStreamCopy.ToArray());
                    }
                    return segment;
                }
```

If you're feeling like trying something, I'm very curious if this would work.   From the docs that you shared on the SB changes, I **_think_** that it should, since we know its a memory stream and has an internal buffer.    If we pass the live suite with this, I'm inclined to take it and push it through our stress tests for a few days to see what happens.

Thoughs?

Reply:
I think the problem is the following: I assume the most likely source here is AmqpMessage. AmqpMessage uses [`MemoryStream(Byte[], Int32, Int32)`](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.-ctor?view=net-6.0#system-io-memorystream-ctor(system-byte()-system-int32-system-int32)) which means `TryGetBuffer` will always return `false`and then we would fall back creating a memory stream with a predefined capacity (of which `TryGetBuffer` would return `true`) only then to write to it and then copy the internal buffer again into the segment because we are calling `ToArray` (which is the unnecessary double allocation this PR gets rid off). 

So what would we gain from this experiment under that promise? Or am I misunderstanding that the sources are most likely AmqpMessages?

What would help is to do a double `TryGetBuffer`. First on the passed stream, then if it returns false allocate the memory stream with the length, copy stuff into it and then `TryGetBuffer` to return it to avoid the `ToArray`. But then again how much value would that add in case we are always getting MemoryStreams managed by AmqpMessage and their internal buffering (still assuming that is indeed the case)?

Reply:
That's where my disconnect is;  the "only works with some constructors".   I think the majority of the time, we'll hit the `BufferListStream` case anyway, coming from `AmqpMessage`, just looking over the [ToStream](https://github.com/Azure/azure-amqp/blob/hotfix/Microsoft.Azure.Amqp/Amqp/AmqpMessage.cs) implementations.    The only other case where we'll hit this method is when someone puts a `Stream` inside the application properties - which is a very rare occurrence.

I agree, not worth the experiment and though the double try-get would guard, unlikely worth the complexity.

Reply:
How about we merge this as is and I spend some more time after my vacation to think about these scenarios in more depth and then send follow up PRs?

Reply:
Agreed; not sure how much more we can optimize and we're hitting the majority cases.   Thank you, once again, for all of your work on this!